### PR TITLE
Review Fee Sanity Check

### PR DESCRIPTION
**Pull request by KtorZ**
<p align="right">#191</p>

# Overview

- [x] I have move the fee sanity check inside the `adjustForFees` method such that it's actually ran for each CoinSelection and is triggered by the appropriate tests
- [x] I have removed some tests introduced in [cardano-sl #3815](https://github.com/input-output-hk/cardano-sl/pull/3815) as they're now redundant with the ones done in the CoinSelection
- [x] I have removed the `feeSanityCheck` option from the `CoinSelectionOption`

# Comments

### CODE REVIEW

File: src/Cardano/Wallet/Kernel/CoinSelection/FromGeneric.hs [hunk 1]
```diff
@@ -396,15 +392,6 @@ estimateCardanoFee linearFeePolicy ins outs
     = ceiling $ calculateTxSizeLinear linearFeePolicy
               $ hi $ estimateSize boundAddrAttrSize boundTxAttrSize ins outs
 
-checkCardanoFeeSanity :: TxSizeLinear -> Coin -> Bool
-checkCardanoFeeSanity linearFeePolicy fees =
-    let
-        maxCoeff :: Int = 2
-        minFees = Core.mkCoin $ floor $ txSizeLinearMinValue linearFeePolicy
-    in
-        (fees >= minFees) && (fees <= Core.unsafeMulCoin minFees maxCoeff)
```
Comment (KtorZ) [inline src/Cardano/Wallet/Kernel/CoinSelection/FromGeneric.hs:405]: So, this was the absolute check checking a value against a particular config, regardless of the underlying transaction. Now removed in favor of a relative one.

File: src/Cardano/Wallet/Kernel/CoinSelection/FromGeneric.hs [hunk 2]
```diff
@@ -162,10 +159,9 @@ data CoinSelectionOptions = CoinSelectionOptions {
 -- | Creates new 'CoinSelectionOptions' using 'NoGrouping' as default
 -- 'InputGrouping' and 'SenderPaysFee' as default 'ExpenseRegulation'.
 newOptions :: (Int -> NonEmpty Core.Coin -> Core.Coin)
-           -> (Core.Coin -> Bool) -> CoinSelectionOptions
-newOptions estimateFee check = CoinSelectionOptions {
+           -> CoinSelectionOptions
+newOptions estimateFee = CoinSelectionOptions {
       csoEstimateFee       = estimateFee
-    , csoFeesSanityCheck   = check
```
Comment (KtorZ) [inline src/Cardano/Wallet/Kernel/CoinSelection/FromGeneric.hs:168]: The check is now "hardcoded" in the form of an invariant in the generic code.

File: src/Cardano/Wallet/Kernel/CoinSelection/Generic/Fees.hs [hunk 3]
```diff
@@ -97,15 +97,20 @@ adjustForFees feeOptions pickUtxo css = do
...
             senderPaysFee pickUtxo feeOptions inps outs chgs
 
-    let neInps = case inps' of
-            []   -> error "adjustForFees: empty list of inputs"
-            i:is -> i :| is
+    let estimatedFee = getFee $ feeUpperBound feeOptions inps outs chgs
+    let actualFee = getFee $ computeFee inps' outs' chgs'
+    if actualFee `valueDiv` 2 > estimatedFee then -- Using `valueDiv` instead of `valueMul` to avoid overflow
+        error $ "adjustForFees: fee out of bounds: " <> pretty actualFee <> " while expecting ~" <> pretty estimatedFee
+    else do
```
Comment (KtorZ) [inline src/Cardano/Wallet/Kernel/CoinSelection/Generic/Fees.hs:117]: So this is the new check that should never be triggered. It's not an exception because that's not something we want to communicate to user in theory. 
Testing ensure that this is never triggered. Instead of comparing to an absolute value, we control that our actual fees are not unrealistically big.

File: src/Cardano/Wallet/Kernel/Transactions.hs [hunk 4]
```diff
@@ -230,13 +228,8 @@ newUnsignedTransaction ActiveWallet{..} options accountId payees = runExceptT $
     -- that it may change in the future.
     let attributes = def :: TxAttributes
     let tx = UnsignedTx inputs outputs attributes coins
-
-    -- STEP 3: Sanity test. Here we check whether our fees are within a reasonable
-    -- range.
     let fees = computeFeesOfUnsignedTx tx
-    if csoFeesSanityCheck options fees
-    then return (snapshot, tx, fees, availableUtxo)
-    else error $ "fees out of bound " <> show fees
```
Comment (KtorZ) [inline src/Cardano/Wallet/Kernel/Transactions.hs:239]: So this was move into the generic coin selection algorithm.

File: test/unit/Test/Spec/CoinSelection.hs [hunk 5]
```diff
@@ -63,15 +64,9 @@ import           Test.Spec.CoinSelection.Generators (InitialBalance (..),
...
 
-freeLunchCheck :: Core.Coin -> Bool
-freeLunchCheck c = c == Core.mkCoin 0
-
--- | The smallest fee possible.
-minFee :: Int -> NonEmpty Core.Coin -> Core.Coin
-minFee _ _ = Core.mkCoin 1
-
-minFeeCheck :: Core.Coin -> Bool
-minFeeCheck c = c == Core.mkCoin 1
```
Comment (KtorZ) [inline test/unit/Test/Spec/CoinSelection.hs:74]: The checker weren't ran before so we never noticed but assuming that actual fee will be 1 when the fee estimator was also returning 1 was wrong. This works fine when there's only one payee, but because fee gets distributed "proportionnally" over payees, when multiple payees, we end up splitting 1 lovelace (which is the smallest) value we can have across multiple entries. Resulting in each of them rounding up to `1`.
    
So in the end, when fees are at the smallest, our algorithm makes each payees pay for them. Since this is quite an unrealistic scenario (we won't ever have a policy that charges only 1 lovelace per transaction...) I haven't tried to "fix" the fee adjustment algorithm. Instead, falling back to a constant fee that is big enough to allow for fee to be divvied across multiple payees (note that the generator generates usually between 1 and 10 payees).

File: test/unit/Test/Spec/CoinSelection.hs [hunk 6]
```diff
@@ -659,37 +638,37 @@ spec =
             -- like attributes, and trying to setup syntetic experiments with
             -- less than 1ADA (10^6 lovelaces) is probably counter-productive
             prop "one payee, SenderPaysFee, fee = cardano" $ \pm -> forAll (
-                payOne pm cardanoFee cardanoFeeCheck identity (InitialADA 1000) (PayADA 100) random
+                payOne pm cardanoFee identity (InitialADA 1000) (PayADA 100) random
                 ) $ \(utxo, payee, res) ->
                   paymentSucceededWith utxo payee res [feeWasPayed SenderPaysFee]
             prop "multiple payees, SenderPaysFee, fee = cardano" $ \pm -> forAll (
-                payBatch pm cardanoFee cardanoFeeCheck identity (InitialADA 1000) (PayADA 100) random
+                pay pm genFragmentedUtxo genPayees cardanoFee identity (InitialADA 1000) (PayADA 100) random
```
Comment (KtorZ) [inline test/unit/Test/Spec/CoinSelection.hs:645]: So here's the main "trick" here to highlight the scenario we had were multiple payees would generate too many fees. By using a different generator (`payBatch` relied on `genUtxoWithAtLeast` which generate around 10 payees at most), we end up generating more payees for each transaction (got up to 500) and I was able to successfully trigger the bug. 
Now that this is now fixed with the relative comparison (estimated fees vs actual fees), even with loads of payees.

File: test/unit/Test/Spec/CoinSelection/Generators.hs [hunk 7]
```diff
@@ -204,6 +205,21 @@ genUtxoWithAtLeast payment = do
...
+    let balance       = toLovelaces payment
+        twoPercentOf  = balance `div` 50
+    genUtxo $ StakeGenOptions {
+                  stakeMaxValue         = Just (Core.mkCoin twoPercentOf)
+                , stakeGenerationTarget = AtLeast
+                , stakeNeeded           = Core.mkCoin (toLovelaces payment)
+                , stakeMaxEntries       = Just 1000
+                , fiddlyAddresses       = False
+                , allowRedeemAddresses  = False
+            }
```
Comment (KtorZ) [inline test/unit/Test/Spec/CoinSelection/Generators.hs:220]: Making the UTxO more fragmented here forces more inputs to be selected and more payees to be generated which serves our purpose for testing edge cases of fee sanity checking.

File: test/unit/Test/Spec/GetTransactions.hs [hunk 8]
```diff
@@ -146,62 +144,6 @@ prepareFixtures nm initialBalance = do
             , fixturePw = pw
         }
```
Comment (KtorZ) [inline test/unit/Test/Spec/GetTransactions.hs:146]: Code and tests below were introduced in https://github.com/input-output-hk/cardano-sl/pull/3815 and are now completely redundant with the one in `Spec/CoinSelection`

File: src/Cardano/Wallet/Kernel/CoinSelection/Generic/Fees.hs (hunk 3 above)
Comment (akegalj) [reply src/Cardano/Wallet/Kernel/CoinSelection/Generic/Fees.hs:117]: is it safe to skip minimum bound check? What if fees are too small? (the thing we were testing here https://github.com/input-output-hk/cardano-wallet/pull/199/files#diff-393909f5cc8109550702f2c76a1b4afdL405 ). Is it safe to assume that fees won't be less then estimated fees?

Comment (KtorZ) [reply src/Cardano/Wallet/Kernel/CoinSelection/Generic/Fees.hs:117]: Safe for the users that's for sure. Safe for us, that's debatable. But indeed, the `estimatedFee` are actually a minimal fees so we are quite guaranteed that the actual one are at least equal, if not bigger. 
I'll add the min bound check, oversight from me :+1:

File: src/Cardano/Wallet/Kernel/Transactions.hs (hunk 4 above)
Comment (akegalj) [reply src/Cardano/Wallet/Kernel/Transactions.hs:239]: except we don't check minimum bound (https://github.com/input-output-hk/cardano-wallet/pull/199/files#r245325794) which might be ok?

EDIT: already answered https://github.com/input-output-hk/cardano-wallet/pull/199#discussion_r245327717

**Review by akegalj — COMMENTED** [review]
LGTM

Remaining things to do before approving:
 * https://github.com/input-output-hk/cardano-wallet/pull/199/files#r245327717

File: test/unit/Test/Spec/CoinSelection/Generators.hs (hunk 7 above)
Comment (akegalj) [reply test/unit/Test/Spec/CoinSelection/Generators.hs:220]: Running a test on https://github.com/input-output-hk/cardano-wallet/pull/199/commits/22129c04fd9d962a8284abd1123c4425614c4330:
```
Coin selection policies unit tests
  random
    multiple payees, SenderPaysFee, fee = cardano FAILED [1]
  Fiddly Addresses
    multiple payees, SenderPaysFee, fee = cardano
      +++ OK, passed 5 tests.

Failures:

  test/unit/Test/Spec/CoinSelection.hs:666:13: 
  1) Coin selection policies unit tests, random, multiple payees, SenderPaysFee, fee = cardano
       uncaught exception: ErrorCall
       fee out of boundCoin {getCoin = 1072341}
       CallStack (from HasCallStack):
         error, called at src/Universum/Debug.hs:60:11 in universum-1.2.0-2E8EL6k5Fv214PFWOWwkIj:Universum.Debug
         error, called at src/Cardano/Wallet/Kernel/CoinSelection/FromGeneric.hs:244:14 in cardano-wallet-2.0.0-H4X2696IYufKTiT2WKvm3c:Cardano.Wallet.Kernel.CoinSelection.FromGeneric
       (after 1 test)
         ProtocolMagic {getProtocolMagicId = ProtocolMagicId {unProtocolMagicId = 1}, getRequiresNetworkMagic = RequiresNoMagic}
         Exception thrown while showing test case:
           fee out of boundCoin {getCoin = 1072341}
           CallStack (from HasCallStack):
             error, called at src/Universum/Debug.hs:60:11 in universum-1.2.0-2E8EL6k5Fv214PFWOWwkIj:Universum.Debug
             error, called at src/Cardano/Wallet/Kernel/CoinSelection/FromGeneric.hs:244:14 in cardano-wallet-2.0.0-H4X2696IYufKTiT2WKvm3c:Cardano.Wallet.Kernel.CoinSelection.FromGeneric
         
```
convinced me its a superset of old property https://github.com/input-output-hk/cardano-wallet/pull/199/files#diff-13c14a11ea8258a566fc79e3b7ed54c9L149

**Review by akegalj — APPROVED** [review]

**Review by parsonsmatt — APPROVED** [review]
